### DISCUSSION TITLE
Terminal renderer

### DISCUSSION THREAD

**Discussion by arcanis** [sourcelair/xterm.js#4980] [Q&A]
I was thinking to use xterm.js to power a multiplexer in a terminal UI library, but I'm not entirely sure what's the best way to achieve that (with "best" being "most correct but simple"). In short, I have to implement a `renderLine` function that, given a line number, returns the ansi sequence to render this line. I did a first attempt by using `@xterm/addon-serialize`, but it cannot run tools like `emacs` properly - I suspect the addon is returning terminal sequences that aren't limited to pure rendering.

My second idea was to use a custom renderer (similar to the canvas / webgl ones), but I looked at the code and they both seem very complex. I'd have expected a way to ask for xterm.js the set of characters and style to apply on every cell from a given range, but I didn't find how to do that just yet. Did I miss some internal API?

**Comment by jerch** [comment]
@arcanis What do you mean with:

> I have to implement a renderLine function that, given a line number, returns the ansi sequence to render this line.

Can you explain more in detail, where the data comes from? How is xterm.js involved here?

My guess currently goes into the direction, that you want to use xterm.js for the multiplexed sub terminals. That should be possible with `xterm-headless` (thus should even work in nodejs).
Idk for sure, if the interface of the serialize addon is fine-grained enough for your output needs, but it is def. the right place to lookup the idea of buffer scraping in xterm.js. This is indeed like an "escape sequence string renderer", still looking up the screen renderers prolly does not help you much here, as those are highly optimized for various aspects hiding the basic scraping logic.

Since you ask for internal APIs - most buffer related things are under [`src/common/buffer`](https://github.com/xtermjs/xterm.js/tree/master/src/common/buffer):
- Buffer.ts - buffer class bundling bufferlines vertically, also deals with reflow, scrollback, attributes etc.
- BufferLine.ts - holds a single row of cell data, has primitives to read&write data
- CellData.ts - convenient class for read access of cell data (used by all renderers, much easier to use than the bufferline primitives)

Reading buffer data down to cell level is also exposed on the official public API (preferable, unless you dont mind hard linking against a certain minor version of xterm.js). Cell writing is not exposed, to see how thats done, see https://github.com/xtermjs/xterm.js/blob/1da0db435f93879a60b9badbb663852ff3223063/src/common/InputHandler.ts#L507-L647.
